pkg/scale/scale.go

```diff
@@ -0,0 +1,93 @@
+package scale
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+)
+
+// package level cache for fieldScaleIndicies
+var cache = &fieldScaleIndicesCache{
+	cache: make(map[string]fieldScaleIndices),
+}
+
+// fieldScaleIndex is used to map field index to scale index
+type fieldScaleIndex struct {
+	fieldIndex int
+	scaleIndex *string
+}
+type fieldScaleIndices []fieldScaleIndex
+
+// fieldScaleIndic
+type fieldScaleIndicesCache struct {
+	cache map[string]fieldScaleIndices
+	sync.RWMutex
+}
+
+func (fsic *fieldScaleIndicesCache) fieldScaleIndices(in interface{}) (v reflect.Value, indices fieldScaleIndices, err error) {
+	t := reflect.TypeOf(in)
+	v = reflect.ValueOf(in)
+	key := fmt.Sprintf("%s.%s", t.PkgPath(), t.Name())
+	if key != "." {
+		var ok bool
+		fsic.RLock()
+		indices, ok = fsic.cache[key]
+		fsic.RUnlock()
+		if ok {
+			return
+		}
+	}
+
+	if !v.IsValid() {
+		err = fmt.Errorf("inputted value is not valid: %v", v)
+		return
+	}
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		tag := field.Tag.Get("scale")
+		switch strings.TrimSpace(tag) {
+		case "":
+			indices = append(indices, fieldScaleIndex{
+				fieldIndex: i,
+			})
+		case "-":
+			// ignore this field
+			continue
+		default:
+			indices = append(indices, fieldScaleIndex{
+				fieldIndex: i,
+				scaleIndex: &tag,
+			})
```

Comment:
can there be a struct where some fields have scale tags and some don't? in that case, is there a chance two fields could end up with the same index? eg
```
type myStruct struct {
    a // add tag for scale index = 1
    b
}
```

both a and b would have index 1

Reply:
they end up getting mapped anyways in the cache, so it will clobber each other depending on the order.  I can throw an err when registering to ensure this doesn't happen.

Reply:
added tests for this.